### PR TITLE
Payments schedule

### PR DESCRIPTION
**Pull request by thomasplevy**
## Description

Refactors methodology for calculating the `date_next_payment` property on a recurring order.

We don't have an open issue for the problem this addresses because we have not been able to adequately come up with a set of recreation steps. However, the problem appears to be that after the *second* payment (the first recurring payment) the stored `date_next_payment` (the date of the transaction that was just processed) is a date in the *future*. As a result of this the calculation is treated like the initial payment calculation and calculated off of the order's `post_date` property.

This is problematic because it results in a new payment being scheduled in the past which is triggered again immediately following the transaction that was just processed.

In this PR we add a fallback condition which will "catch" this scenario and calculate the next payment off of the last transaction date.

## How has this been tested?

There are a ton of new integration tests which attempt to simulate the above issue as well as adding more robust test coverage for recurring payment calculations in "good" conditions.

I've have manually run recurring transactions from the frontend of the website (the way a customer/student would initiate a transaction). And then manually adjusted the next payment date to ensure that the recurring transaction fires as expected when triggered by the background processes.

I also have hundreds of recurring charges that run regularly on my local. I'll review my logs tomorrow to ensure they all continue to charge and reschedule as expected.

## Types of changes

Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] My code has been tested.
- [x] My code passes all existing automated tests. 
- [x] My code follows the LifterLMS Coding & Documentation Standards.

### CODE REVIEW

File: includes/models/model.llms.order.php [hunk 1]
```diff
@@ -317,9 +337,18 @@ private function calculate_next_payment_date( $format = 'Y-m-d H:i:s' ) {
 		if ( 0 != $end_time && ( $next_payment_time + 23 * HOUR_IN_SECONDS ) > $end_time ) {
 			$ret = '';
 		} elseif ( $next_payment_time > 0 ) {
-			$ret = date_i18n( $format, $next_payment_time );
+			$ret = date( $format, $next_payment_time );
```
Comment (eri-trabiccolo) [inline includes/models/model.llms.order.php]: what's the reason for this change? Is this intended? Just a question.

**Comment by eri-trabiccolo** [comment]
@thomasplevy about:

> However, the problem appears to be that after the second payment (the first recurring payment) the stored date_next_payment (the date of the transaction that was just processed) is a date in the future. As a result of this the calculation is treated like the initial payment calculation and calculated off of the order's post_date property.

Not sure about this, I mean the `date_next_payment` set in the future can justify transactions duplicated several times in a range of 10 minutes?
You know, looking at the code, a failure of `current_time()` seems to fit more, although not sure how that could happen.

Anyways the fall-back on the last transaction date as starting "time" for the next payment date should solve the issue, so we should be fine.

File: includes/models/model.llms.order.php (hunk 1 above)
Comment (thomasplevy) [reply includes/models/model.llms.order.php]: My thinking is that this is an internal/private function that we request a specific format for so we actually don't want any modifications to that requested format adjusted by `date_i18n()`

Comment (eri-trabiccolo) [reply includes/models/model.llms.order.php]: I see, makes sense, thanks.
